beefy-gadget/src/gossip.rs

```diff
@@ -53,41 +49,48 @@ where
 /// rejected/expired.
 ///
 ///All messaging is handled in a single BEEFY global topic.
-pub(crate) struct BeefyGossipValidator<B>
+pub(crate) struct GossipValidator<B>
 where
 	B: Block,
 {
 	topic: B::Hash,
 	live_rounds: RwLock<Vec<NumberFor<B>>>,
 }
 
-impl<B> BeefyGossipValidator<B>
+impl<B> GossipValidator<B>
 where
 	B: Block,
 {
-	pub fn new() -> BeefyGossipValidator<B> {
-		BeefyGossipValidator {
+	pub fn new() -> GossipValidator<B> {
+		GossipValidator {
 			topic: topic::<B>(),
 			live_rounds: RwLock::new(Vec::new()),
 		}
 	}
 
-	/// Note a live voting round
+	/// Note a voting round.
+	///
+	/// Noting `round` will keep `round` live.
 	///
-	/// This should be called, in order to keep tabs on `round`.
+	/// We retain the [`MAX_LIVE_GOSSIP_ROUNDS`] most **recent** voting rounds as live.
+	/// As long as a voting round is live, it will be gossiped to peer nodes.
 	pub(crate) fn note_round(&self, round: NumberFor<B>) {
 		trace!(target: "beefy", "🥩 About to note round #{}", round);
 
-		let mut live_rounds = self.live_rounds.write();
+		let mut live = self.live_rounds.write();
 
-		// NOTE: ideally we'd use a VecDeque here, but currently binary search is only available on
-		// nightly for `VecDeque`.
-		while live_rounds.len() > MAX_LIVE_GOSSIP_ROUNDS {
-			let _ = live_rounds.remove(0);
+		// we want to keep `MAX_LIVE_GOSSIP_ROUNDS` most recent rounds.
+		if (live.len() == MAX_LIVE_GOSSIP_ROUNDS) && (live[0] > round) {
+			return;
 		}
 
-		if let Some(idx) = live_rounds.binary_search(&round).err() {
-			live_rounds.insert(idx, round);
+		// we want to keep `MAX_LIVE_GOSSIP_ROUNDS` *after* noting `round`
+		while live.len() > (MAX_LIVE_GOSSIP_ROUNDS - 1) {
+			let _ = live.remove(0);
+		}
+
+		if let Some(idx) = live.binary_search(&round).err() {
```

Comment:
```suggestion
		if let Some(idx) = live.binary_search(&round).err() {
			live.insert(idx, round);
		}
		if|while live.len() > MAX_LIVE_GOSSIP_ROUNDS {
			let _ = live.remove(0);
		}
```

I think this is something that is going to be more future proof and is just simpler. In the previous code you still have a case where we remove a round, but then we don't really add one (because it's already there).